crates/storage/provider/src/providers/blockchain_provider.rs

```diff
@@ -3357,4 +3336,960 @@ mod tests {
 
         Ok(())
     }
+
+    #[test]
+    fn transaction_id_found_in_database() -> eyre::Result<()> {
+        let mut rng = generators::rng();
+        let (provider, database_blocks, _, _) = provider_with_random_blocks(
+            &mut rng,
+            TEST_BLOCKS_COUNT,
+            0,
+            BlockRangeParams {
+                tx_count: TEST_TRANSACTIONS_COUNT..TEST_TRANSACTIONS_COUNT,
+                ..Default::default()
+            },
+        )?;
+
+        // Choose a random transaction from the database blocks
+        let tx = &database_blocks[0].body[0];
+        let tx_hash = tx.hash();
+
+        println!("tx hash: {:?}", tx);
+
+        println!("result: {:?}", provider.transaction_id(tx_hash)?);
+
+        // Ensure the transaction ID can be found in the database
+        let result = provider.transaction_id(tx_hash)?;
+        assert_eq!(result, Some(0));
+
+        Ok(())
+    }
+
+    #[test]
+    fn transaction_id_found_in_memory() -> eyre::Result<()> {
+        let mut rng = generators::rng();
+        let (provider, _, in_memory_blocks, _) = provider_with_random_blocks(
+            &mut rng,
+            TEST_BLOCKS_COUNT,
+            TEST_BLOCKS_COUNT,
+            BlockRangeParams {
+                tx_count: TEST_TRANSACTIONS_COUNT..TEST_TRANSACTIONS_COUNT,
+                ..Default::default()
+            },
+        )?;
+
+        // Choose a random transaction from the in-memory blocks
+        let tx = &in_memory_blocks[0].body[0];
+        let tx_hash = tx.hash();
+
+        // Ensure the transaction ID can be found in the in-memory state
+        let result = provider.transaction_id(tx_hash)?;
+        assert!(result.is_some(), "Transaction ID should be found in the in-memory state");
+
+        // Check that the transaction ID is greater than the last database transaction ID
+        let last_db_tx_id = provider.database.last_block_number()?;
+        let last_db_tx_id =
+            provider.database.block_body_indices(last_db_tx_id)?.unwrap().last_tx_num();
+
+        assert!(
+            result.unwrap() > last_db_tx_id,
+            "In-memory transaction ID should be greater than the last database transaction ID"
+        );
+        assert_eq!(result, Some(20));
```

Comment:
can we rely on `last_db_tx_id + 1` here instead of a hardcoded value?

Reply:
fixed